### PR TITLE
Working and tested nextcloud role

### PR DESCRIPTION
**Pull request by Scriptkiddi**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+| `nextcloud_version`             | `13.0.0`                  | Version number of the Nextcloud version to be installed                                                                           |
+| `nextcloud_system_user`         | `www-data`                | User under which Nextcloud should run. The user has to exist.                                                                     |
+| `nextcloud_system_group`        | `www-data`                | Group under which Nextcloud should run. The group has to exist.                                                                   |
+| `nextcloud_path`                | `/var/www/nextcloud`      | Path to install nextcloud to                                                                                                      |
+| `nextcloud_path_old`            | `/var/www/nextcloud_old`  | Path to store the nextcloud dir when upgrading.                                                                                   |
+| `nextcloud_data_dir_path`       | `/var/www/nextcloud_data` | Path to store the nextcloud data in. This can not be in the `{{ nextcloud_path }}` its self, because we move dirs during upgrades |
```
Comment (SlothOfAnarchy) [inline README.md]: *itself

File: README.md [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+| `nextcloud_plugins`             | `[]`                      | List of Nextcloud plugins to install and activate                                                                                 |
+| `nextcloud_config`              | `[]`                      | Dict of Nextcloud setting read more below                                                                                         |
+| `nextcloud_ldap_enable`         | `false`                   | Enable ldap                                                                                                                       |
+| `nextcloud_ldap`                | `false`                   | Dict containing the ldap attributes                                                                                               |
+
+### LDAP Settings 
+This is a list of the ldap attributes for a description and behaviour please see the [Nextcloud documentation](https://docs.nextcloud.com/server/13/admin_manual/configuration_user/user_auth_ldap.html#configuration)
```
Comment (SlothOfAnarchy) [inline README.md]: comma

File: README.md [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+To set a LDAP attribute put the attribute in the nextcloud_ldap dict.
+```yml
+nextcloud_ldap:
+  ldapHost: ldaps://ldap.example.com
+```
+
+| Name                                      |
```
Comment (SlothOfAnarchy) [inline README.md]: Is there a column missing? Why isn't this a list?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+| `ldapUuidUserAttribute`         |
+| `turnOffCertCheck`              |
+| `turnOnPasswordChange`          |
+| `useMemberOfToDetectMembership` |
+
+
+### `nextcloud_config`
+`nextcloud_config` is an object containing the settings for the Nextcloud instance. Do not use it to enable/disable apps and/or for setting up ldap.
+The setup of this object is similar to the output of the `php occ config:list` option.
+Under nextcloud_config you should have two objects called `system` and `apps` under those objects you can further objects to describe the configuration as an example setting the trusted domains for Nextcloud.
```
Comment (SlothOfAnarchy) [inline README.md]: block quotes

Comment (SlothOfAnarchy) [inline README.md]: > you can further objects

?

File: scripts/nextcloud.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        return "integer"
+    elif isinstance(value, float):
+        return "float"
+    return "string"
+
+
+def generate_paths(data):
+    """ Return list of lists containing the required values to pass to call """
+    execute_list = []
+    if not isinstance(data, list) and not isinstance(data, dict):
```
Comment (SlothOfAnarchy) [inline scripts/nextcloud.py]: why the second condition? Are dicts also being detected as lists in python?

File: tasks/configure.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+---
+- name: Create tmp file
+  tempfile:
+    state: file
+  register: tmp_file
+  changed_when: False
+
+- name: Set owner for tmp file
```
Comment (SlothOfAnarchy) [inline tasks/configure.yml]: why is this necessary?

**Review by SlothOfAnarchy — CHANGES_REQUESTED** [review]
more to follow

File: scripts/nextcloud.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        plain = ['sudo',
+                 '-u',
+                 USER,
+                 'php',
+                 'occ',
+                 'config:{0}:set'.format(region),
+                 '--value={0}'.format(value_to_set),
+                 '--type={0}'.format(value_type)]
+        plain[6:6] = command[1:-1]
+        plain = plain if command[0] == "system" else plain[:-1]
```
Comment (SlothOfAnarchy) [inline scripts/nextcloud.py:67]: Why isn't the type necessary for app parameters?

File: tasks/download.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  delegate_to: localhost
+
+- name: Extract Nextcloud into tmp dir
+  unarchive:
+    src: "{{ global_cache_dir }}/nextcloud-{{ nextcloud_download_version }}.zip"
+    dest: "/tmp"
+    owner: "{{ nextcloud_system_user }}"
+    group: "{{ nextcloud_system_user }}"
+
+- name: Copy Nextcloud from tmp dir to install dir
```
Comment (SlothOfAnarchy) [inline tasks/download.yml:16]: Why not extract it directly to the specified path?

File: tasks/install.yml [hunk 9]
```diff
@@ -0,0 +1,24 @@
+---
+- include_tasks: download.yml
+  vars:
```
Comment (SlothOfAnarchy) [inline tasks/install.yml:3]: nice

File: tasks/download.yml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  command: "mv -f /tmp/nextcloud {{ nextcloud_extract_path }}"
+
+- name: Set permissions on root
+  file: 
+    path: "{{ nextcloud_extract_path }}"
+    state: directory 
+    mode: 0755
+    owner: "{{ nextcloud_system_user }}"
```
Comment (SlothOfAnarchy) [inline tasks/download.yml:24]: What about the group?

File: tasks/install.yml [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    nextcloud_download_version: "{{ nextcloud_version }}"
+    nextcloud_extract_path: "{{ nextcloud_path }}"
+
+- name: Create a new database with name '{{ nextcloud_mysql_database_name }}'
+  mysql_db:
+    name: "{{ nextcloud_mysql_database_name }}"
+    state: present
```
Comment (SlothOfAnarchy) [inline tasks/install.yml]: Isn't this the default?

File: tasks/install.yml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: Create mariadb user
+  mysql_user:
+    name: "{{ nextcloud_mysql_user }}"
+    password: "{{ nextcloud_mysql_password }}"
+    priv: '{{ nextcloud_mysql_database_name }}.*:ALL,GRANT'
+    state: present
```
Comment (SlothOfAnarchy) [inline tasks/install.yml]: Isn't this the default?

File: tasks/install.yml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: Run the nextcloud installer from command line
+  shell: 'sudo -u {{ nextcloud_system_user }} php occ  maintenance:install --database "mysql" --database-name "{{nextcloud_mysql_database_name}}"  --database-user "{{nextcloud_mysql_user}}" --database-pass "{{nextcloud_mysql_password}}" --admin-user "{{ nextcloud_user }}" --admin-pass "{{ nextcloud_password }}" --data-dir "{{ nextcloud_datadir_path }}"'
+  args:
+    chdir: "{{ nextcloud_path }}/"
+
+
```
Comment (SlothOfAnarchy) [inline tasks/install.yml]: remove this line

File: tasks/install.yml [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  mysql_user:
+    name: "{{ nextcloud_mysql_user }}"
+    password: "{{ nextcloud_mysql_password }}"
+    priv: '{{ nextcloud_mysql_database_name }}.*:ALL,GRANT'
+    state: present
+
+- name: Run the nextcloud installer from command line
+  shell: 'sudo -u {{ nextcloud_system_user }} php occ  maintenance:install --database "mysql" --database-name "{{nextcloud_mysql_database_name}}"  --database-user "{{nextcloud_mysql_user}}" --database-pass "{{nextcloud_mysql_password}}" --admin-user "{{ nextcloud_user }}" --admin-pass "{{ nextcloud_password }}" --data-dir "{{ nextcloud_datadir_path }}"'
+  args:
+    chdir: "{{ nextcloud_path }}/"
```
Comment (SlothOfAnarchy) [inline tasks/install.yml]: I'm sure there is a filter for the directory enforcing.

File: tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,42 @@
+---
+- name: Update cache and install packages
+  apt:
+    name: "{{item}}"
+    update_cache: yes
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: This task will be changed if only the cache changes. We found out about a different solution in the past.

File: tasks/ldap.yml [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+- name: Get LDAP config id
+  shell: "sudo -u {{ nextcloud_system_user }} php occ ldap:show-config | grep -w 'Configuration' | awk -F'|' '{print $3}'"
+  args:
+    chdir: "{{ nextcloud_path }}"
+  register: config_id_exists_output
+  changed_when: false
+
+- name: Get config id from stdout
+  set_fact:
+    config_id: "{{ config_id_exists_output.stdout | trim  }}"
```
Comment (SlothOfAnarchy) [inline tasks/ldap.yml]: space

File: tasks/ldap.yml [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  when: config_id_exists_output.stdout != ""
+  changed_when: false
+
+- name: Create empty LDAP config
+  shell: "sudo -u {{ nextcloud_system_user }} php occ ldap:create-empty-config -p"
```
Comment (SlothOfAnarchy) [inline tasks/ldap.yml]: Why `sudo` instead of `become_user`? I thought we enabled SSH pipelining to circumvent the become ansible bug?

File: tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    - php-mbstring
+    - php-ldap
+    - python-mysqldb
+    - php-mysql
+    - php-xml
+    - python-yaml
+    - php-redis
+
+- name: Check if Nextcloud folder exists
+  stat: path={{nextcloud_path}}
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: indent

File: tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+  register: st
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: What does `st` stand for?

File: tasks/plugins.yml [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+  register: result
+  failed_when: result.rc != 0 and not "already installed" in result.stdout
+  changed_when: not "already installed" in result.stdout
+
+- name: Activate plugins
+  shell: "sudo -u www-data php occ app:enable -n {{ item }}"
+  args:
+    chdir: "{{ nextcloud_path }}"
+  with_items: "{{ nextcloud_plugins }}"
+  register: result
```
Comment (SlothOfAnarchy) [inline tasks/plugins.yml]: `result` is not used.
Also, why not change this task?

File: tasks/update.yml [hunk 21]
```diff
@@ -0,0 +1,53 @@
+---
+- name: Enable maintenance mode
+  shell: "sudo -u {{ nextcloud_system_user }} php occ maintenance:mode --on"
+  args:
+    chdir: "{{ nextcloud_path }}"
+
+- name: MySQL and Redis
```
Comment (SlothOfAnarchy) [inline tasks/update.yml]: Stop MySQL, Nginx and Redis

File: tasks/update.yml [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+  systemd:
+    name: "{{ item }}"
+    state: stopped
+  with_items:
+    - nginx
+    - redis
+    - mysql
+
+- name: "Move Nextcloud dir to {{ nextcloud_path_old }}"
```
Comment (SlothOfAnarchy) [inline tasks/update.yml]: What if `nextcloud_path_old` already exists or contains old data?
Maybe autmatically name the folder depending on the version number.

File: tasks/update.yml [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+- name: Move apps over to the new install
+  command: "mv {{ nextcloud_path_old }}/apps {{ nextcloud_path }}/apps"
+
+- name: Run Nextcloud upgrade
+  shell: "sudo -u {{ nextcloud_system_user }} php occ upgrade"
+  args:
+    chdir: "{{ nextcloud_path }}"
+
+- name: Start Nginx MySQL and Redis
```
Comment (SlothOfAnarchy) [inline tasks/update.yml]: ,

**Review by SlothOfAnarchy — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Scriptkiddi) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (Scriptkiddi) [reply README.md]: fixed

File: README.md (hunk 3 above)
Comment (Scriptkiddi) [reply README.md]: Because I found a table nicer to look at

File: README.md (hunk 4 above)
Comment (Scriptkiddi) [reply README.md]: fixed

Comment (Scriptkiddi) [reply README.md]: fixed

File: scripts/nextcloud.py (hunk 5 above)
Comment (Scriptkiddi) [reply scripts/nextcloud.py]: because when you get something else than a dict or a list you want to return just the data

File: tasks/configure.yml (hunk 6 above)
Comment (Scriptkiddi) [reply tasks/configure.yml]: removed

File: scripts/nextcloud.py (hunk 7 above)
Comment (Scriptkiddi) [reply scripts/nextcloud.py:67]: because type is only a valid parameter for system values

File: tasks/download.yml (hunk 8 above)
Comment (Scriptkiddi) [reply tasks/download.yml:16]: because i want to remvoe one directory. when you extract the zip you get /tmp/nextcloud/* and if you want to name your folder something else

File: tasks/download.yml (hunk 10 above)
Comment (Scriptkiddi) [reply tasks/download.yml:24]: fixed

File: tasks/install.yml (hunk 11 above)
Comment (Scriptkiddi) [reply tasks/install.yml]: fixed

File: tasks/main.yml (hunk 19 above)
Comment (Scriptkiddi) [reply tasks/main.yml]: renamed to result

File: tasks/plugins.yml (hunk 20 above)
Comment (Scriptkiddi) [reply tasks/plugins.yml]: yes it is see when condition on update nextcloud

File: tasks/update.yml (hunk 21 above)
Comment (Scriptkiddi) [reply tasks/update.yml]: fixed

File: tasks/update.yml (hunk 23 above)
Comment (Scriptkiddi) [reply tasks/update.yml]: fixed

File: tasks/update.yml (hunk 22 above)
Comment (Scriptkiddi) [reply tasks/update.yml]: fixed

**Comment by Scriptkiddi** [comment]
@SlothOfAnarchy please rereview

File: scripts/nextcloud.py (hunk 5 above)
Comment (SlothOfAnarchy) [reply scripts/nextcloud.py]: Okay, I overlooked the first `not`.

**Comment by SlothOfAnarchy** [comment]
@Scriptkiddi fine so far, but you missed a few comments in my second review.

File: tasks/ldap.yml (hunk 17 above)
Comment (Scriptkiddi) [reply tasks/ldap.yml]: because become is a shit solution and no we did it so we have no problem switching users

File: tasks/main.yml (hunk 18 above)
Comment (Scriptkiddi) [reply tasks/main.yml]: fixed

File: tasks/ldap.yml (hunk 16 above)
Comment (Scriptkiddi) [reply tasks/ldap.yml]: fixed

File: tasks/install.yml (hunk 13 above)
Comment (Scriptkiddi) [reply tasks/install.yml]: fixed

File: tasks/main.yml (hunk 15 above)
Comment (Scriptkiddi) [reply tasks/main.yml]: ?

File: tasks/install.yml (hunk 14 above)
Comment (Scriptkiddi) [reply tasks/install.yml]: fixed

**Comment by Scriptkiddi** [comment]
@SlothOfAnarchy pls rereview

File: tasks/main.yml (hunk 15 above)
Comment (SlothOfAnarchy) [reply tasks/main.yml]: Use a separate task for updating cache like in all our roles...
Also the spaces :P

File: tasks/ldap.yml (hunk 17 above)
Comment (SlothOfAnarchy) [reply tasks/ldap.yml]: Why is become a shit solution? It is intended exactly for tasks like this, becoming a specific user.

File: tasks/install.yml (hunk 12 above)
Comment (SlothOfAnarchy) [reply tasks/install.yml]: still

File: tasks/plugins.yml (hunk 20 above)
Comment (SlothOfAnarchy) [reply tasks/plugins.yml]: No this is another variable that is coincidentally also named `result`.

**Review by SlothOfAnarchy — CHANGES_REQUESTED** [review]
See my comments.

File: tasks/plugins.yml (hunk 20 above)
Comment (Scriptkiddi) [reply tasks/plugins.yml]: fixed

File: tasks/main.yml (hunk 15 above)
Comment (Scriptkiddi) [reply tasks/main.yml]: fixed

**Comment by Scriptkiddi** [comment]
@SlothOfAnarchy please rereview

File: tasks/main.yml [hunk 24]
```diff
@@ -1,9 +1,13 @@
 ---
-- name: Update cache and install packages
+- name: Update cache
   apt:
-    name: "{{item}}"
     update_cache: yes
   when: ansible_pkg_mgr == "apt"
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: `changed_when: false`

**Review by SlothOfAnarchy — CHANGES_REQUESTED** [review]

**Comment by Scriptkiddi** [comment]
@SlothOfAnarchy fixed

**Review by SlothOfAnarchy — APPROVED** [review]
